rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+use directly. To achieve the aim of making tests no harder to write
+than SpecialPowers-based Gecko tests, there is also a higher-level API
+initially providing two main capabilities: the ability to
+`postMessage` a remote context, and the ability to `executeScript` so
+that the script runs in a remote context.
+
+This API is provided by a `RemoteWindow` object. The `RemoteWindow`
+object doesn't handle creating the browsing context, but given the
+UUID for the remote window, creates a `SendChannel` which is able to
+send messages to the remote. Alternatively the `RemoteWindow` may be
+created first and its `uuid` property used when constructing the URL.
+
+Inside the remote browsing context itself, the test author has to call
+`await start_window()` in order to set up a `RecvChannel` with UUID given
+by the `uuid` parameter in `location.href`. The returned object offers
+an `addMessageHandler(callback)` API to receive messages sent with the
+`postMessage` API on `RemoteWindow`, and `async nextMessage()` to wait
+for the next message.
+
+##### Script Execution
+
+`RemoteWindow.executeScript(fn, ...args)` serializes the function
+`fn`, using `Function.toString()`. Each argument is
+serialized using the remote value serialization algorithm. Along with
+the function string and the arguments, a `SendChannel` is sent for the
+command response (only one such channel is created per `RemoteWindow`
+for efficiency reasons), and a command id is sent to disambiguate
+responses from different commands.
+
+On the remote side the function is deserialized and executed. If
+execution results in a `Promise` value, the result of that promise is
+awaited. The final return value after the promise is resolved is sent
+back and forms the async return value of the `executeScript` call. If
+the script throws, the thrown value is provided as the result, and
+re-thrown in the originating context. In addition an `exceptionDetails`
+field provides the line/column numbers of the original exception,
+where available.
+
+In addition there is a `RemoteWindow.executeScriptNoResult(fn,
+...args)` method. This works the same way except no channel is passed,
+and so no result is returned. This can be useful in case the script
+does something like trigger a navigation, so there's no need to
+synchronize the navigation starting (which will close the socket) with
+writing the response.
+
+TODO: the naming here isn't great. In particular a `RemoteWindow`
+could actually be some other kind of global like a worker, and
+`start_window()` is a pretty nondescript method name.
+
+#### Navigation and bfcache
+
+For specific use cases around bfcache, it's important to be able to
+ensure that no network connections (including websockets) remain open
+at the time of navigation, otherwise the page will be excluded from
+bfcache. In the current prototype this is handled as follows:
+
+* A `pause` method on `SendChannel`. This causes a server-initiated
```

Comment:
As commented at https://github.com/web-platform-tests/wpt/pull/29803#discussion_r707551472, I'm not sure how `pause` works and whether `pause` is needed (the draft impl doesn't use `pause`).

Reply:
So the problem pause is intended to solve is where you have something like:
```
await remote.executeScript(() => startNavigation());
await remote.executeScript(() => ensureNavigationHappened());
```
With a `poll()` based interface this can work as written; the remote simply doesn't call `poll()` after `startNavigation` and before the navigation is complete. But with websockets, we don't have any synchronisation between the navigation and sending the `ensureNavigationHappened` message. If the test sends the second message before the socket is closed on the remote, it can end up being lost. So pause solves this by effectively providing a synchronisation point around the socket being closed:
```
await remote.executeScript(() => startNavigation());
remote.pause();
// We now know the remote now won't listen for messages until it reconnects to the socket after navigation
await remote.executeScript(() => ensureNavigationHappened());
```

From the point of view of the remote, `closeAllSockets()` and `pause()` do race, but the effect of `pause()` is a server-initiated disconnect of the read channel for the remote, and the effect of `closeAllSockets` is a client-initiated disconnect of all channel-related sockets, so the ordering isn't very important.

We can avoid requiring pause by not sending requests to a remote after initiating a navigation, before getting a confirmation that the navigation is actually complete. In practice this means that the target page has to be a different `RemoteWindow` (i.e. have a different uuid) and has to process a message i.e. something like:
```
await remote.executeScript(() => startNavigation());
await navigationTargetRemote.executeScript(() => goBack());
await remote.executeScript(() => ensureNavigationHappened());
```
That's what the bfcache tests are doing now, and why `pause` isn't needed there, but either way the ergonomics are a bit tricky. I think we could drop it from the initial featureset, but these races around navigation are tricky to understand and debug.